src/plat/pc99/machine/dev/vga/io.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2024, DornerWorks
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ */
+#include <config.h>
+#include <arch/kernel/boot_sys.h>
+#include <arch/model/statedata.h>
+#include <machine/io.h>
+#include <plat/machine/io.h>
+#include <drivers/uart.h>
+
+#define VGA_PHYS 0xb8000 // Text buffer address
+
+//black background, light grey text, no blinking
+#define VGA_COLOR 0x07
+#define CLEAR_SCREEN_CHAR 32
+static uint64_t vga_addr = 0;
+static uint64_t vga_size = 0;
+// assume 80x25 console
+#define MAX_X_POS 80
+#define MAX_Y_POS 25
+
+/* Name kept 'serial_init' since other code expects to call this */
+void serial_init(uint16_t port)
+{
+    set_vga_addr(VGA_PHYS);
+}
+
+void set_vga_addr(uint64_t addr)
+{
+    vga_addr = addr;
+    // multiply by 2 since each slot in the VGA buffer has a char byte and 'style' byte
+    vga_size = MAX_X_POS * MAX_Y_POS * 2;
+}
+
+uint64_t get_vga_addr(void)
+{
+    return vga_addr;
+}
+
+static int verify_vga_offset(uint64_t offset)
+{
+    if (offset > vga_size) {
+        return -1;
+    }
+    return 0;
+}
+
+static void vga_write(uint16_t data, uint64_t offset)
+{
+    if (!verify_vga_offset(offset)) {
```

Comment:
Since we are reading and writing `uint16`, we should only be writing to offsets that are divisible by 2, otherwise we'd be writing the content field of one char and the colour field of the next. We could also potentially be reading/writing one byte past the screen area. So we should probably make this something like:

```c
    if (offset < vga_size && (offset & 1 == 0)) {
      *(uint16_t *)(vga_addr + offset) = data;
    }
```

Then again, if this function is only used inside `vga_putchar` we could also not check, because we do always multiply by 2 in there. In that case, adding a comment that we expect offset to be divisible by 2 here would be good.